keps/sig-auth/20190916-noderestriction-pods.md

```diff
@@ -0,0 +1,309 @@
+updates _cannot_ modify the OwnerReferences [[1][]].
+
+The second mechanism is through the creation of "mirror pods". The Kubelet can run pods from other
+sources than the API server, such as a static manifest directory or pulled from an HTTP
+server. These pods are referred to as "static pods". Since static pods don't come from the API
+server, components that read pods from the API wouldn't know about them. To compensate, the Kubelet
+creates a "mirror pod", which reflects the state of the static pod it's running.
+
+Mirror pods have some special properties. They are identified with a special annotation,
+`kubernetes.io/config.mirror`, and the Kubelet is only authorized to create mirror pods, and only on
+the same node (itself). The Kubelet won't run a mirror pod (since it's actually running a static
+pod). Mirror pods are also restricted from using a service account, or secrets, configmaps,
+persistent volumes, and other resources restricted by the node authorizer, in order to prevent an
+attacker from bypassing the authorization by creating mirror pods.
+
+[1]: https://github.com/kubernetes/kubernetes/blob/ab73a018de51bddf9d03d6fed6e867b60196c796/pkg/registry/core/pod/strategy.go#L162-L171
+
+[Node Authorizer]: https://kubernetes.io/docs/reference/access-authn-authz/node/
+[NodeRestriction]: https://kubernetes.io/docs/reference/access-authn-authz/node/
+[man-in-the-middle]: https://en.wikipedia.org/wiki/Man-in-the-middle_attack
+
+### Threat Model
+
+At a high level, this proposal targets clusters making effective use of node isolation to separate
+sensitive workloads or limit the blast radius of a successful node compromise. More specifically, it
+makes the following assumptions:
+
+- The cluster uses scheduling constraints such as node selectors & taints / tolerations to separate
+  workloads of different trust or privilege levels.
+- An attacker has compromised a low-privilege node meaning they have code execution as root in the
+  host namespaces (i.e. a container escape). Low-privilege in this case means the node is not
+  hosting any privileged workloads with permissions to trivially take over the cluster
+  (e.g. unrestricted pod creation, read secrets, etc).
+
+### Goals
+
+- Prevent a compromised node from manipulating controllers to execute confused deputy attacks.
+
+### Non-Goals
+
+- Solving all node isolation issues.
+- Solving all possible man-in-the-middle attacks.
+
+## Proposal
+
+All restrictions will be enforced through the [NodeRestriction][] admission controller. These
+extensions will be guarded by the `NodeRestrictionPods` feature gate.
+
+On update, only the delta will be restricted. In other words, the Kubelet can modify pods with
+restricted labels / annotations / owner references as long as it's not modifying (or adding or
+deleting) the restricted entries.
+
+### Label Restrictions
+
+The Kubelet will be prevented from updating pod labels or creating mirror pods with labels, except
+for whitelisted keys:
```

Comment:
This means that user-namespaced labels `example.com/foo` would no longer be permitted on static pods, which doesn't seem to match our [traditional posture of leaving all non-k8s-namespaced labels up to the end user](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/):
> Labels are intended to be used to specify identifying attributes of objects that are meaningful and relevant to users, but do not directly imply semantics to the core system.

The [enforced restrictions on node labels](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction) specifically only limit k8s-namespaced labels, since those are under the project's control.

Were alternatives considered, like (for example):
* providing a way for the cluster administrator to indicate specific labels are permitted in mirror pods
* making mirror pods be opt-in for selection by workload controllers / services / endpoints

Reply:
The case for preventing modification of non-static-pod labels is clearer to me, since the kubelet does not do that currently

Reply:
Addressed these under "alternatives considered", but briefly:

> providing a way for the cluster administrator to indicate specific labels are permitted in mirror pods

This is what the whitelisted prefix is for. We could provide a configurable option, but I'm not convinced it's worth it.

> making mirror pods be opt-in for selection by workload controllers / services / endpoints

Since selection happens on the controller side, every controller needs to be opted in. This distributed control might not be implemented by third-party controllers, making this option more error-prone.

Reply:
Also, to address your first point:

> Labels are intended to be used to specify identifying attributes of objects that are meaningful and relevant to users

IMO static pods should be considered system workloads, not user workloads. Under that condition, all labels on static pods are system labels.

Reply:
>> providing a way for the cluster administrator to indicate specific labels are permitted in mirror pods
>
> This is what the whitelisted prefix is for. We could provide a configurable option, but I'm not convinced it's worth it.

This is similar to the debate we had in node label limitation. I don't think we can limit labels outside our control (non-k8s.io labels) unconditionally. For node labels, we only put unconditional policy around labels within our own namespace.

>> making mirror pods be opt-in for selection by workload controllers / services / endpoints
>
> Since selection happens on the controller side, every controller needs to be opted in. This distributed control might not be implemented by third-party controllers, making this option more error-prone.

fair

> > Labels are intended to be used to specify identifying attributes of objects that are meaningful and relevant to users
> 
> IMO static pods should be considered system workloads, not user workloads. Under that condition, all labels on static pods are system labels.

There are three groups I'm considering:
* kubernetes component developers (use kube-system, k8s.io-prefixed labels, etc)
* cluster administrators (use arbitrary labels to set up their cluster topology)
* non-cluster administrators (can use arbitrary labels to set up their application topology)

A cluster admin (today) has complete control over labels outside the k8s.io/kubernetes.io namespace. To start denying them the ability to use those labels seems problematic.